File: src/experience_buffer.rs
use rand::distributions::Distribution;
use rand::{rngs::SmallRng, SeedableRng};
use tch::{Device, IndexOp, Kind, Scalar, Tensor};
use torch_sys::IntList;
pub struct ExperienceStats {
    means: Tensor,
    msqs: Tensor,
    count: Tensor,
}

impl ExperienceStats {
    pub fn new(shape: impl IntList + Copy, device: Device) -> Self {
        Self {
            means: Tensor::zeros(shape, (Kind::Float, device)),
            msqs: Tensor::ones(shape, (Kind::Float, device)),
            count: Tensor::zeros(1, (Kind::Int64, device)),
        }
    }

    pub fn push(&mut self, value: &Tensor) {
        self.count += 1;
        let delta = value - &self.means;
        self.means += &delta / &self.count;
        let delta2 = value - &self.means;
        self.msqs += delta * delta2;
    }

    pub fn mean(&self) -> &Tensor {
        &self.means
    }

    pub fn var(&self) -> Tensor {
        &self.msqs / Scalar::float((self.count.int64_value(&[0]) - 1) as f64)
    }
}

pub struct RandomExperienceBuffer {
    obs_size: i64,
    pub curr_states: Tensor,
    curr_actions: Tensor,
    rewards: Tensor,
    pub next_states: Tensor,
    dones: Tensor,
    size: i64,
    next_idx: i64,
    capacity: i64,
    minsize: i64,
    rng: SmallRng,
    device: Device,
    pub stats: ExperienceStats,
    normalize_obs: bool,
}

impl RandomExperienceBuffer {
    pub fn new(
        capacity: i64,
        obs_size: i64,
        minsize: i64,
        seed: u64,
        normalize_obs: bool,
        device: Device,
    ) -> Self {
        Self {
            obs_size,
            curr_states: Tensor::empty([capacity, obs_size], (Kind::Float, device)),
            curr_actions: Tensor::empty(capacity, (Kind::Int64, device)),
            rewards: Tensor::empty(capacity, (Kind::Float, device)),
            next_states: Tensor::empty([capacity, obs_size], (Kind::Float, device)),
            dones: Tensor::empty(capacity, (Kind::Int8, device)),
            capacity,
            next_idx: 0,
            size: 0,
            minsize,
            rng: SmallRng::seed_from_u64(seed),
            device,
            stats: ExperienceStats::new(obs_size, device),
            normalize_obs,
        }
    }

    pub fn ready(&self) -> bool {
        self.size >= self.minsize
    }

    pub fn add(
        &mut self,
        curr_state: &Tensor,
        curr_action: usize,
        reward: f32,
        done: bool,
        next_state: &Tensor,
    ) {
        let index: i64 = self.obs_size * self.next_idx;
        let index = Vec::from_iter(index..(index + self.obs_size));
        let index = &Tensor::from_slice(&index).to_device(self.device);

        let curr_state = &curr_state.to_device(self.device);
        let curr_action = &Tensor::from(curr_action as i64).to_device(self.device);
        let reward = &Tensor::from(reward).to_device(self.device);
        let done = &Tensor::from(done as i8).to_device(self.device);
        let next_state = &next_state.to_device(self.device);

        self.curr_states = self.curr_states.put(index, curr_state, false);
        self.next_states = self.next_states.put(index, next_state, false);

        let index = &Tensor::from(self.next_idx).to_device(self.device);

        self.curr_actions = self.curr_actions.put(index, curr_action, false);
        self.rewards = self.rewards.put(index, reward, false);
        self.dones = self.dones.put(index, done, false);

        self.next_idx = (self.next_idx + 1) % self.capacity;
        self.size = self.capacity.min(self.size + 1);

        if self.normalize_obs {
            self.stats.push(curr_state);
        }
    }

    pub fn normalize(&self, values: Tensor) -> Tensor {
        // let (var, mean) = self.curr_states.var_mean_dim(0, false, false);
        // ((values - mean) / var.sqrt()).to_device(self.device)
        if self.normalize_obs {
            (values - self.stats.mean()) / (self.stats.var()).sqrt()
        } else {
            values
        }
    }

    pub fn sample_batch(&mut self, size: usize) -> (Tensor, Tensor, Tensor, Tensor, Tensor) {
        let dist: rand::distributions::Uniform<i64> =
            rand::distributions::Uniform::new(0i64, self.size as i64);
        let index: Vec<i64> = dist.sample_iter(&mut self.rng).take(size).collect();
        // println!("{index:?}");
        (
            self.normalize(self.curr_states.i(index.clone()))
                .to_kind(Kind::Float),
            self.curr_actions.i(index.clone()).reshape([-1, 1]),
            self.rewards.i(index.clone()).reshape([-1, 1]),
            self.dones.i(index.clone()).reshape([-1, 1]),
            self.normalize(self.next_states.i(index))
                .to_kind(Kind::Float),
        )
    }
}

File: python-src/experience_buffer.py
import torch
from small_rng_rs import SmallRng
import numpy as np
from torch import tensor


class ExperienceStats:
    def __init__(self, shape, device):
        self.means = torch.zeros(shape, dtype=torch.float64, device=device)
        self.msqs = torch.ones(shape, dtype=torch.float64, device=device)
        self.count = torch.zeros(1, dtype=torch.int64, device=device)

    def push(self, value):
        self.count += 1
        delta = value - self.means
        self.means += delta / self.count
        delta2 = value - self.means
        self.msqs += delta * delta2

    def mean(self):
        return self.means

    def var(self):
        return self.msqs / (self.count.item() - 1)


class RandomExperienceBuffer:
    def __init__(self, capacity, obs_size, minsize, seed, normalize_obs, device):
        self.obs_size = obs_size
        self.curr_states = torch.empty(
            (capacity, obs_size), dtype=torch.float64, device=device
        )
        self.curr_actions = torch.empty(capacity, dtype=torch.int64, device=device)
        self.rewards = torch.empty(capacity, dtype=torch.float64, device=device)
        self.next_states = torch.empty(
            (capacity, obs_size), dtype=torch.float64, device=device
        )
        self.dones = torch.empty(capacity, dtype=torch.int8, device=device)
        self.capacity = capacity
        self.next_idx = 0
        self.size = 0
        self.minsize = minsize
        self.rng = SmallRng(seed)
        self.device = device
        self.stats = ExperienceStats(obs_size, device)
        self.normalize_obs = normalize_obs

    def ready(self):
        return self.size >= self.minsize

    def add(self, curr_state, curr_action, reward, done, next_state):
        curr_state = curr_state.to(self.device)
        curr_action = tensor(curr_action, dtype=torch.int64, device=self.device)
        reward = tensor(reward, dtype=torch.float64, device=self.device)
        done = tensor(done, dtype=torch.int8, device=self.device)
        next_state = next_state.to(self.device)

        self.curr_states[self.next_idx] = curr_state
        self.next_states[self.next_idx] = next_state

        self.curr_actions[self.next_idx] = curr_action
        self.rewards[self.next_idx] = reward
        self.dones[self.next_idx] = done

        self.next_idx = (self.next_idx + 1) % self.capacity
        self.size = min(self.capacity, self.size + 1)

        if self.normalize_obs:
            self.stats.push(curr_state)

    def normalize(self, values):
        if self.normalize_obs:
            return (values - self.stats.mean()) / torch.sqrt(self.stats.var())
        else:
            return values

    def sample_batch(self, size):
        indices = self.rng.integers(0, self.size, size)
        indices = tensor(indices, dtype=torch.int64, device=self.device)
        return (
            self.normalize(self.curr_states[indices]).float(),
            self.curr_actions[indices].reshape(-1, 1),
            self.rewards[indices].reshape(-1, 1),
            self.dones[indices].reshape(-1, 1),
            self.normalize(self.next_states[indices]).float(),
        )
